File: weekly-contest/weekly-contest-160/2-circular-permutation-in-binary-representation/-backtrack.js
/**
 * 回溯
 */
var circularPermutation = function (n, start) {
  const LENGTH = 1 << n
  let res

  function backtrack (curr, acc, visit) {
    if (acc.length === LENGTH) {
      const tmp = acc[0] ^ acc[acc.length - 1]
      if (!(tmp & tmp - 1)) {
        res = acc
        return true
      }
      return false
    }

    let mask = 1
    for (let i = 0; i < n; ++i) {
      const next = curr ^ mask
      if (!visit.has(next)) {
        visit.add(next)
        acc.push(next)
        if (backtrack(next, acc, visit)) return true
        // 回溯
        visit.delete(next)
        acc.pop()
      }
      mask = mask << 1
    }
  }

  backtrack(start, [start], new Set([start]))
  return res
};

module.exports = circularPermutation

File: problems/386-lexicographical-numbers/backtrack.cpp
/**
 * 回溯（DFS能保证字典序）
 * 
 * 时间：`O()`, 8ms
 */

class Solution {
public:
    vector<int> lexicalOrder(int n) {
        N = n;
        // 最高位不能为0，只能是1～9
        for (int i = 1; i <= min(9, n); ++i) {
            backtrack(i);
        }
        return res;
    }

    void backtrack(int curr) {
        res.push_back(curr);
        // 之后的位置上可以是0～9
        for (int i = 0; i <= 9; ++i) {
            int next = curr * 10 + i;
            if (next > N) break; // 剪枝
            backtrack(next);
        }
    }

private:
    int N;
    vector<int> res;
};